crates/matrix-sdk/src/widget_api/driver/widget.rs

```diff
@@ -0,0 +1,78 @@
+use async_trait::async_trait;
+
+use super::handler::Result;
+use crate::widget_api::messages::capabilities;
+
+/// The Widget trait that needs to be implemented in the native client.
+/// An instance implementing this trait will be required to initialize the widget driver.
+#[async_trait]
+pub trait Widget {
+    /// This function will be called whenever there is a new message available
+    /// that has to be sent to the widget.
+    ///
+    /// # Examples
+    /// ```
+    /// implement Widget for MyWidget {
+    ///     fn send_widget_message(&self, message: &str) -> Result<()> {
+    ///         myWidgetIFrame.postmessage(message);
+    ///         Ok()
+    ///     }
+    /// }
+    ///```
+    fn send_widget_message(&self, message: &str) -> Result<()>;
+
+    /// The client should show a dialog to give the user the option approve some/all of the capabilites that the widget requests.
+    /// The returned value contains the capability options the user has approved.
+    /// The client should also provide good phrasing for the different permissions/filters:
+    /// cap.send_room_event = [{event_type: "m.room.message", msgtype: "m.image" }] should be come sth like:
+    /// Allow the widget to, send images in this room. (checkbox)
+    async fn capability_permissions(
+        &self,
+        cap: capabilities::Options,
+    ) -> Result<capabilities::Options>;
+
+    /// The client should show a dialog to approve if the widget is allowed to get an OpenId token.
+    /// A `OpenIdDialogResponse` is returned containing a flag if the user alloed the token request and
+    /// if the permission can be cached so the widget driver will not call `openid_permissions` on the widget again.
+    async fn openid_permissions(&self) -> Result<OpenIdDialogResponse>;
+
+    /// Returns the widget id from the widget state event.
+    fn id(&self) -> &str;
+
+    /// Returns the widget state event as a raw string.
+    fn get_widget_state_json(&self) -> &str;
```

Comment:
we maybe also want to do that inside the rust sdk...
We need to discuss the library design here.
 - either we could just get the id from the native side and than during the initialization fetch the state event and parse it so that we know if the widget initializes immediately or if we should wait for ContentLoaded (this information is inside the room state that defines the room widget)
  - Or we let the native code take care of this and pass us the initalization data (wait for I frame load or whatever name it has) with the widget info.